src/stateManager/StateManager.ts

```diff
@@ -1459,9 +1588,129 @@ class StateManager {
         };
     }
 
+    private getPreviousStateSnapshotOrThrow(
+        coordinates: BlockCoordinates
+    ): StateSnapshot {
+        const previousStateSnapshot =
+            this.storage.getPreviousStateSnapshot(coordinates);
+        if (!previousStateSnapshot)
+            throw new Error(
+                "createStateSnapshot for block - previousStateSnapshot undefined"
+            );
+        return previousStateSnapshot;
+    }
+
+    private getPendingInboundMessageBlocks(
+        previousStateSnapshot: StateSnapshot
+    ): MessageBlockStruct[] {
+        const latestStoredHash =
+            this.storage.inboundMessages.getLatestBlockHash();
+        if (!latestStoredHash) {
+            return [];
+        }
+
+        const previousHash =
+            previousStateSnapshot.snapshotData.latestInboundMessageBlockHash;
+
+        if (previousHash && latestStoredHash === previousHash) {
+            return [];
+        }
+
+        return this.storage.inboundMessages.getMessageBlocksInRange(
+            latestStoredHash,
+            previousHash ?? ethers.ZeroHash
+        );
+    }
+
+    private findBrokenInboundMessageChainBlock(
+        previousStateSnapshot: StateSnapshot,
+        inboundMessageBlocks: MessageBlockStruct[]
+    ): MessageBlockStruct | undefined {
+        if (inboundMessageBlocks.length === 0) {
+            return undefined;
+        }
+
+        let expectedPreviousHash =
+            previousStateSnapshot.snapshotData.latestInboundMessageBlockHash ??
+            ethers.ZeroHash;
+        let expectedHeight = BigInt(
+            previousStateSnapshot.snapshotData
+                .latestInboundMessageBlockHeight ?? 0n
+        );
+
+        for (const inboundBlock of inboundMessageBlocks) {
+            if (inboundBlock.previousBlockHash !== expectedPreviousHash) {
+                return inboundBlock;
+            }
+            expectedHeight += 1n;
+            if (BigInt(inboundBlock.blockHeight ?? 0n) !== expectedHeight) {
+                return inboundBlock;
+            }
+            expectedPreviousHash = hash(
+                Codec.encode(inboundBlock, Type.MessageBlock)
+            );
+        }
+
+        return undefined;
+    }
+
+    private async applyInboundMessageBlocksToState(
+        inboundMessageBlocks: MessageBlockStruct[],
+        totalDeposits: BalanceStruct,
+        encodedState?: Bytes
+    ): Promise<{ encodedState: Bytes; totalDeposits: BalanceStruct }> {
+        let updatedTotalDeposits = totalDeposits;
+
+        if (inboundMessageBlocks.length === 0) {
+            return {
+                encodedState:
+                    encodedState ?? (await this.diamondStateMachine.getState()),
+                totalDeposits: updatedTotalDeposits
+            };
+        }
+
+        for (const messageBlock of inboundMessageBlocks) {
+            for (const message of messageBlock.messages) {
+                const processed =
```

Comment:
do i understand correctly?
processInboundMessage will process the `data` field
and the later `addBalance is the uniform accounting for all message types